src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnectionCache.scala

```diff
@@ -0,0 +1,520 @@
+            config.bulkConfig.maxThroughputForBulkOperations.get,
+            maxAvailableThroughput
+          )
+        } else {
+          maxAvailableThroughput
+        }
+
+      val builder = DocumentBulkExecutor.builder
+        .from(
+          client,
+          config.database,
+          config.container,
+          pkDef,
+          effectivelyAvailableThroughputForBulkOperations
+        )
+        .withInitializationRetryOptions(bulkExecutorInitializationRetryOptions)
+        .withMaxUpdateMiniBatchCount(config.bulkConfig.maxMiniBatchUpdateCount)
+        .withMaxMiniBatchSize(config.bulkConfig.maxMiniBatchImportSizeKB * 1024)
+
+      // Instantiate DocumentBulkExecutor
+      val bulkExecutor = builder.build()
+
+      attemptClientCacheEntryUpdate(
+        config,
+        clientCacheEntry,
+        operationName = "BulkExecutor",
+        oldClientCacheEntry => oldClientCacheEntry.copy(bulkExecutor = Some(bulkExecutor))
+      )
+
+      bulkExecutor
+    }
+  }
+
+  def getPartitionKeyDefinition(config: ClientConfiguration): PartitionKeyDefinition = {
+    config.bulkConfig.partitionKeyOption match {
+      case None =>
+        val containerMetadata: ContainerMetadata = getOrReadContainerMetadata(
+          config
+        )
+        containerMetadata.partitionKeyDefinition
+      case Some(pk) =>
+        val pkDefinition = new PartitionKeyDefinition()
+        val paths: ListBuffer[String] = new ListBuffer[String]()
+        paths.add(pk)
+        pkDefinition.setPaths(paths)
+        pkDefinition
+    }
+  }
+
+  /**
+    * Resets the partition key range cache within the Document client
+    * This is done to enforce that the DocumentClient retrieves fresh partition key range
+    * metadata - for example when computing the number of partitions to execute
+    * queries on etc.
+    *
+    * @param config The cache key - represents the config settings
+    */
+  def reinitializeClient(config: ClientConfiguration): Unit = {
+    val clientCacheEntry: ClientCacheEntry = clientCache.get(config)
+    if (clientCacheEntry != null) {
+      val containerMetadata: ContainerMetadata = getOrReadContainerMetadata(config)
+      val docClient: DocumentClient = clientCacheEntry.docClient
+      logDebug(
+        s"Updating Partition key range cache for DocumentClient#${docClient.hashCode()} " +
+          s"of ClientConfiguration#${config.hashCode()}"
+      )
+      val field = docClient.getClass.getDeclaredField("partitionKeyRangeCache")
+      field.setAccessible(true)
+      val partitionKeyRangeCache =
+        field.get(docClient).asInstanceOf[PartitionKeyRangeCache]
+      val range =
+        new com.microsoft.azure.documentdb.internal.routing.Range[String](
+          "00",
+          "FF",
+          true,
+          true
+        )
+
+      partitionKeyRangeCache.getOverlappingRanges(
+        containerMetadata.selfLink,
+        range,
+        true
+      )
+
+      attemptClientCacheEntryUpdate(
```

Comment:
It may not be necessary to update the cache entry. My impression as a C# dev reading this is that we have a ref to the docClient instance in the cache and that modifying it would be sufficient and the pointer to it doesn't need to be updated however using reflection this way in Scala is outside my experience.

Reply:
Yes you are right - because DocumentClient is a "reference type" and not a case class "value type" functionally it works. The reason why I added the cache-refresh was that my impression was that many Scala Developers really hate when case classes are mutated - even indirectly by changing the content of a reference type. Anyway - given that it isn't technically necessary and we won't win a beauty price for this code anyway I'll remove it.